node/src/exchange/client.rs

```diff
@@ -58,13 +76,79 @@ impl ExchangeClientHandler {
 
         let req_id = req_resp.send_request(&peer, request);
         self.reqs.insert(req_id, ReqInfo { amount, respond_to });
+    }
 
-        trace!("Request initiated");
+    fn send_head_request(
+        &mut self,
+        req_resp: &mut ReqRespBehaviour,
+        request: HeaderRequest,
+        respond_to: OneshotResultSender<Vec<ExtendedHeader>, P2pError>,
+    ) {
+        const MAX_PEERS: usize = 10;
+        const MIN_HEAD_RESPONSES: usize = 2;
+
+        let Some(peers) = self.peer_tracker.best_n_peers(MAX_PEERS) else {
+            respond_to.maybe_send_err(P2pError::NoPeers);
+            return;
+        };
+
+        let mut rxs = Vec::with_capacity(peers.len());
+
+        for peer in peers {
+            let (tx, rx) = oneshot::channel();
+
+            let req_id = req_resp.send_request(&peer, request.clone());
+            let req_info = ReqInfo {
+                amount: 1,
+                respond_to: tx,
+            };
+
+            self.reqs.insert(req_id, req_info);
+            rxs.push(rx);
+        }
+
+        // Choose the best HEAD.
+        //
+        // Algorithm: https://github.com/celestiaorg/go-header/blob/e50090545cc7e049d2f965d2b5c773eaa4a2c0b2/p2p/exchange.go#L357-L381
+        spawn(async move {
+            let mut resps: Vec<_> = join_all(rxs)
+                .await
+                .into_iter()
+                .filter_map(|v| v.ok()?.ok()?.into_iter().next())
+                .collect();
+            let mut counter: HashMap<_, usize> = HashMap::new();
+
+            // In case of no responses, Celestia handles it as NotFound
+            if resps.is_empty() {
+                respond_to.maybe_send_err(ExchangeError::HeaderNotFound);
+                return;
+            }
+
+            // Sort by height in descending order
+            resps.sort_by_key(|resp| Reverse(resp.height()));
+
+            // Count peers per response
+            for resp in &resps {
+                *counter.entry(resp.hash()).or_default() += 1;
+            }
+
+            // Return the header with the maximum height that was received by at least 2 peers
+            for resp in &resps {
+                if counter[&resp.hash()] >= MIN_HEAD_RESPONSES {
+                    respond_to.maybe_send_ok(vec![resp.to_owned()]);
+                    return;
+                }
+            }
+
+            // Otherwise return the header with the maxium height
+            respond_to.maybe_send_ok(vec![resps[0].to_owned()]);
+        });
     }
 
     #[instrument(level = "trace", skip(self, responses), fields(responses.len = responses.len()))]
     pub(super) fn on_response_received(
         &mut self,
+        peer: PeerId,
```

Comment:
this peer_id is tricky as it is only used by the `instrument` macro. wdyt about moving it to the end. prefixing with an underscore and renaming back to `peer` in instrument macro?

Reply:
applies to all places

Reply:
For now it is used only instrument. But the reason I added it was for `PeerTracker` scoring later on.

Reply:
yeah, I just started scanning the code to look why is it needed and it's not really obvious that instrument uses that, so wanted to indicate that fact before we implement the actual usage, but it's a nitpick anyway so up to you